packages/network-controller/src/NetworkController.ts

```diff
@@ -136,6 +160,46 @@ function isInfuraProviderType(type: string): type is InfuraNetworkType {
   return Object.keys(InfuraNetworkType).includes(type);
 }
 
+/**
+ * Builds an identifier for an Infura network client, used to look it up in the
+ * registry at some later point in time.
+ *
+ * @param infuraNetworkName - The name of the Infura network.
+ * @returns The built identifier.
+ */
+function buildInfuraNetworkClientId(
+  infuraNetworkName: InfuraNetworkType,
+): InfuraNetworkClientId {
+  return `infura||${infuraNetworkName}` as const;
+}
+
+/**
+ * Builds an identifier for a custom network client, used to look it up in the
+ * registry at some later point in time.
+ *
+ * @param networkConfigurationId - The ID of the network configuration for the
+ * network.
+ * @param chainId - The chain ID of the network.
+ * @param rpcUrl - The RPC URL of the network.
+ * @returns The built identifier.
+ */
+function buildCustomNetworkClientId(
+  networkConfigurationId: NetworkConfigurationId | undefined,
+  chainId?: Hex,
+  rpcUrl?: string | undefined,
+): CustomNetworkClientId {
+  if (networkConfigurationId === undefined) {
+    if (chainId === undefined) {
+      throw new Error('Missing chainId (this should never happen)');
+    }
+    if (rpcUrl === undefined) {
```

Comment:
These two branches both seem unreachable. If there is no RPC URL, how would this client function?

Reply:
You can call this function one of two ways, but I admit that that isn't obvious. I've made a change here, hopefully it makes things more clear: [3c7879f](https://github.com/MetaMask/core/pull/1439/commits/3c7879fc130066bdb0083e4dbe696554bad1cdbd)

Reply:
Sorry I'm not following. Could you give an example where the rpcUrl is expected to be undefined? This still seems unreachable to me. All custom networks have RPC URLS, it's the only part that's completely necessary for them to function.

Reply:
Even the updated description you added indicate that this is unreachable ("This always has an RPC URL," re: the second argument type), though confusingly the type still shows it as optional.

Reply:
I went back and made it so that you can just pass a provider config if you like, that way I don't have to explain anything in the JSDocs: [ec75958](https://github.com/MetaMask/core/pull/1439/commits/ec75958bfb66617501ac1acf872f60ee44792809)

Reply:
I extracted the validation step which checks for `chainId` and `rpcUrl` to a new function `validateCustomProviderConfig`. This function types the provider config appropriately, so now by the time that we call `buildCustomNetworkId`, we already know that we have a provider config that has an RPC URL, so we can skip the check there. Should make this look a little less gross. ([24aa829](https://github.com/MetaMask/core/pull/1439/commits/24aa8295a11d062b7cb24ee933f867bd9955163c))

Reply:
>However, in the type this property is optional.

Ah, I missed that part. Hadn't realize we didn't know it was set at the type level by this point. OK I'll take another look.